### ISSUE TITLE
Parent dictionary item in Umbraco dictionary item datasource changes in every environment

### ISSUE THREAD

**Issue by simonech** [leekelleher/umbraco-contentment#345]
### Which Contentment version are you using?

4.4.4

### Which Umbraco version are you using? For example: 8.14.1 - don't just write v8

12.0

### Bug summary

We store all our datatypes using uSync and synchronise dictionary items using uSync publisher between environments.
But when we deploy and the datatypes are imported into new environments, the Parent dictionary item is not kept, since it contains the reference to the dictionary item id of the instance of the developer that did it the first time.
So after each deploy we need to update the datasource again with the correct dictionary item.

I guess the reason is that the dictionary parent is referenced via the id, so the id changes with each environment, unlike guids that are kept the same or XPath that do not rely on any identifier.

I suspect this issue might also affects Umbraco deploy on Umbraco cloud.

### Steps to reproduce

- create a dictionary item and sync via usync publisher
- create a data source, set the parent item and serialise via usync
- import the same usync file in both environment

### Expected result / actual result

I'd expect the datasource to still show the correct parent item, like it happens with root nodes specified for content pickers.

What I get now is a non-working data source since it has no parent item specified

### Do you have Umbraco ModelsBuilder enabled?

- [ ] Yes, it is enabled.

**Comment by leekelleher** [comment]
Thanks @simonech . I can't recall if Dictionary items have a Guid or not. I'll investigate.

If they don't, then it'll be a case of writing a uSync value-converter/mapper (whatever they are called), to handle the reference at export/import times.

**Comment by simonech** [comment]
Yeah, if dictionaries have guids that would the best... otherwise... would be feature request for Umbraco 😄 
I didn't know of such thing as usync mappers at import... I'll have a look at those, thx

**Comment by leekelleher** [comment]
Just had a look, they are called `ISyncMapper`, but limited on examples in the uSync repo: https://github.com/KevinJump/uSync/tree/v12/main/uSync.Community.Contrib/Mappers

**Comment by leekelleher** [comment]
Good news, the dictionary items do have a Guid.  Bad news is that Umbraco's built-in treepicker component, doesn't return the Guid/key for Dictionary items... Contentment is using the treepicker, so at that point it only has access to the `int` ID value.

I recall at the time checking the documentation to see if I could use the AngularJS `dictionaryResource` to get the Guid/key, but [according to the docs](https://our.umbraco.com/apidocs/v8/ui/#/api/umbraco.resources.dictionaryResource), there was nothing to get info about the dictionary item, (and it felt overkill to start writing my own API controller for a little used end point) ... also if Umbraco fixes the underlying issue, then Contentment would leverage that.

Anyway, looking at this again today, I've noticed that the `dictionaryResource` does actually have a `getById` method, and has for a long time... _(Contentment supports from v8.17.0),_ turns out that the ngdocs comments for the method are incorrect, so it never showed up in the documentation! 🤦 https://github.com/umbraco/Umbraco-CMS/blob/release-8.17.0/src/Umbraco.Web.UI.Client/src/common/resources/dictionary.resource.js#L89

So, I can add code to look up the dictionary item's Guid/key upon selection.
> here: https://github.com/leekelleher/umbraco-contentment/blob/4.4.6/src/Umbraco.Community.Contentment/DataEditors/DictionaryPicker/dictionary-picker.js#L51

In terms of backwards-compatibility, I'll need to add code to check if the configured Parent Dictionary Item is either an `int` or `Guid`, but that's no problem.

I'm not sure when I'll get chance to do this, (mostly as the Umbraco Community Teams visit will take up most of next week). It might be worth pursuing the `ISyncMapper` approach in the interim?  But in the end, I definitely prefer that Contentment stores the `Guid` instead of the `int`.

**Comment by juanlondonoforero** [comment]
Hi @leekelleher ; first of all thanks for such a great tool very useful
I'm adding the comment here because it's very similar and I wonder if maybe it can be covered on the same fix/release
In my case I'm not using uSync and I'm just simply deploying via umbraco Cloud but I'm also loosing the connection between the Contentment property and the Parent Dictionary item. I even tried to export the Dictionary as content but still it didn't work. 
do you have any ideas on how to fix this?
many thanks

**Comment by leekelleher** [comment]
@juanlondonoforero Thanks for reporting this. Yes, it's the same issue and would be resolved by the same fix. I know how to fix it, just need dev time to do it, (it'll be in the next couple of weeks).

**Comment by juanlondonoforero** [comment]
Thanks Lee. I'm sorry if I sound needy but would it be possible to know what tables/fields on the DB I need to manually change to make it work (even if it's not an elegant fix) ; The client added all the content on staging and I know that next time we deploy from DEV to Staging the content for all the Contentment fields will be lost even if I re-connect the data source and they would have to re-enter them one by one. 
many thanks in advanced

**Comment by leekelleher** [comment]
@juanlondonoforero The underlying issue is that between environments the `int` ID for the dictionary items are different, whereas the `Guid` key is consistent.  The fix (in Contentment) is to use the `Guid`, but I wont have dev time available this week to do so, (due to other Umbraco Community Team commitments).  

I can't offer any advise on what to manually change at database level, as that'd be the same as manually re-configurating the Data List's data-source in the backoffice.

The only workable suggestion for Umbraco Cloud, would be to try tackle something similar to uSync's mappers, using Umbraco Deploy's [`IDataTypeConfigurationConnector` interface](https://github.com/umbraco/Umbraco-CMS/blob/release-12.1.2/src/Umbraco.Core/Deploy/IDataTypeConfigurationConnector.cs).  This would enable you to write custom code to convert the `int` to a `Guid` for deployment, then convert it back at the end. See the `ToArtifact` and `FromArtifact` methods. I can't offer any examples of this code at the moment.

Failing that, if you need something immediately, it might be better to write a custom data-source to replace the "Umbraco Dictionary Items" data-source? Using a hard-coded `Guid` key of the parent dictionary item?
Feel free to reuse whatever code is needed from here: https://github.com/leekelleher/umbraco-contentment/blob/4.4.6/src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/UmbracoDictionaryDataListSource.cs#L64

**Comment by simonech** [comment]
@leekelleher do you have a release date for 4.5?

**Comment by leekelleher** [comment]
@simonech I'm aiming to get the v4.5.0 release out this week.
(I'd intended for last week, but I caught the dreaded flu, wiped me out, gah 🤧)

**Comment by simonech** [comment]
@leekelleher Thank you!
